scripts/lib/file

```diff
@@ -0,0 +1,238 @@
+#!/usr/bin/env bash
+
+# This script contains functions to create OLM and registry files.
+
+# create_catalogsource_file creates a CatalogSource for the OLM.
+function create_catalogsource_file() {
+  local catalogsource_file="$1"
+  local op_name="$2"
+  local op_image="$3"
+  local namespace="$4"
+
+  cat <<EOF > "$catalogsource_file"
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: ${op_name}-ocs
+  namespace: $namespace
+spec:
+  displayName: $op_name Community Operator
+  image: $op_image
+  publisher: community-operators
+  sourceType: grpc
+EOF
+}
+
+# create_subscription_file creates a subscription for OLM to manage.
+function create_subscription_file() {
+  local subscription_file="$1"
+  local op_name="$2"
+  local pkg_name="$3"
+  local csv_name="$4"
+  local namespace="$5"
+
+  cat <<EOF > "$subscription_file"
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: ${op_name}-sub
+  namespace: $namespace
+spec:
+  name: $pkg_name
+  source: ${op_name}-ocs
+  sourceNamespace: $namespace
+  startingCSV: $csv_name
+  channel: alpha
+EOF
+}
+
+# create_operator_group_file creates an operator group for the operator CSV.
+function create_operator_group_file() {
+  local op_group_file="$1"
+  local name="$2"
+  local service_account="$3"
+  local namespace="$4"
+
+  cat <<EOF > "$op_group_file"
+apiVersion: operators.coreos.com/v1alpha2
+kind: OperatorGroup
+metadata:
+  name: ${name}-og
+  namespace: $namespace
+spec:
+  serviceaccount: $service_account
+  targetNamespaces:
+  - $namespace
+EOF
+}
+
+# create_registry_dockerfile creates a registry Dockerfile.
+function create_registry_dockerfile() {
+  local reg_dockerfile="$1"
+  local op_name="$2"
+
+  cat <<EOF > "$reg_dockerfile"
+FROM quay.io/openshift/origin-operator-registry:latest
+COPY $op_name manifests/${op_name}
+RUN /usr/bin/initializer -o bundles.db
+ENTRYPOINT ["/usr/bin/registry-server"]
+CMD ["-d", "bundles.db", "-t", "termination.log"]
+EOF
+}
+
+# create_cr_files_from_metadata takes a CSV's metadata bytes and transforms CR's
+# from the "alm-examples" annontation, if present, into YAML manifest files.
+function create_cr_files_from_metadata() {
+  local csv_file="$1"
+  local cr_dir="$2"
+  local namespace="$3"
+
+  local alm_examples="$(yq r "$csv_file" "metadata.annotations.alm-examples")"
+  [[ "$alm_examples" == "null" ]] && return 0
+
+  local tmp
+  while read -r obj; do
+    tmp="$(mktemp -p "$cr_dir" "XXXXXXXXXX.cr.yaml")"
+    echo "$obj" | yq w - "metadata.namespace" "$namespace" > "$tmp"
+  done < <(echo "$alm_examples" | jq -c '.[]')
+}
+
+function create_permissions_files() {
+  local deploy_dir="$1"
+  local csv_file="$2"
+  local op_name="$3"
+  local namespace="$4"
+
+  declare -A service_accounts_set
+  local prefix sa_name
+
+  local permissions_json="$(yq r --tojson "$csv_file" "spec.install.spec.permissions")"
```

Comment:
If the operator is installed into the AllNamespaces operatorGroup, permissions get upgraded to clusterPermissions by OLM.
@alecmerdler Please confirm.

Reply:
This is correct (https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/olm/operatorgroup.go#L297).

Reply:
For now operators are only tested with the SingleNamespace OperatorGroup, so this isn't an issue. However, for operators that only support AllNamespaces, this will need to be revisited.